### PR TITLE
#19358 Avoid to run the ReindexQueueAPI#addStructureReindexEntries

### LINKED ISSUE #19358: Large amounts of content in structure can deadlock the DB when repermissioning
If you are running a large dotCMS installation and you repermission a content type or host, we mark all the content records for reindex.  We do an "insert into" to fill the reindex table.  Our query here needs some performance tuning as this can deadlock a db if the amount of content is too large.

```
	at org.apache.tomcat.jdbc.pool.StatementFacade$StatementProxy.invoke(StatementFacade.java:114)
	at com.sun.proxy.$Proxy43.execute(Unknown Source)
	at com.dotmarketing.common.db.DotConnect.executeQuery(DotConnect.java:609)
	at com.dotmarketing.common.db.DotConnect.executeQuery(DotConnect.java:522)
	at com.dotmarketing.common.db.DotConnect.loadResult(DotConnect.java:261)
	at com.dotmarketing.common.reindex.ReindexQueueFactory.addStructureReindexEntries(ReindexQueueFactory.java:116)
	at com.dotmarketing.common.reindex.ReindexQueueAPIImpl.addStructureReindexEntries_aroundBody0(ReindexQueueAPIImpl.java:47)
	at com.dotmarketing.common.reindex.ReindexQueueAPIImpl$AjcClosure1.run(ReindexQueueAPIImpl.java:1)
	at org.aspectj.runtime.reflect.JoinPointImpl.proceed(JoinPointImpl.java:149)
	at com.dotcms.aspects.aspectj.AspectJDelegateMethodInvocation.proceed(AspectJDelegateMethodInvocation.java:42)
	at com.dotcms.aspects.interceptors.WrapInTransactionMethodInterceptor$$Lambda$31/803979942.execute(Unknown Source)
	at com.dotmarketing.db.LocalTransaction.wrapReturnWithListeners(LocalTransaction.java:59)
	at com.dotcms.aspects.interceptors.WrapInTransactionMethodInterceptor.invoke(WrapInTransactionMethodInterceptor.java:23)
	at com.dotcms.aspects.aspectj.WrapInTransactionAspect.invoke(WrapInTransactionAspect.java:41)
	at com.dotmarketing.common.reindex.ReindexQueueAPIImpl.addStructureReindexEntries(ReindexQueueAPIImpl.java:46)
	at com.dotmarketing.business.PermissionBitFactoryImpl.resetPermissionReferences(PermissionBitFactoryImpl.java:2912)
	at com.dotmarketing.business.PermissionBitFactoryImpl.savePermission(PermissionBitFactoryImpl.java:1555)
	at com.dotmarketing.business.PermissionBitFactoryImpl.assignPermissions(PermissionBitFactoryImpl.java:1534)
	at com.dotmarketing.business.PermissionBitAPIImpl.assignPermissions_aroundBody24(PermissionBitAPIImpl.java:664)
	at com.dotmarketing.business.PermissionBitAPIImpl$AjcClosure25.run(PermissionBitAPIImpl.java:1)
	at org.aspectj.runtime.reflect.JoinPointImpl.proceed(JoinPointImpl.java:149)
	at com.dotcms.aspects.aspectj.AspectJDelegateMethodInvocation.proceed(AspectJDelegateMethodInvocation.java:42)
	at com.dotcms.aspects.interceptors.WrapInTransactionMethodInterceptor$$Lambda$31/803979942.execute(Unknown Source)
	at com.dotmarketing.db.LocalTransaction.wrapReturnWithListeners(LocalTransaction.java:59)
	at com.dotcms.aspects.interceptors.WrapInTransactionMethodInterceptor.invoke(WrapInTransactionMethodInterceptor.java:23)
	at com.dotcms.aspects.aspectj.WrapInTransactionAspect.invoke(WrapInTransactionAspect.java:41)
	at com.dotmarketing.business.PermissionBitAPIImpl.assignPermissions(PermissionBitAPIImpl.java:618)
	at com.dotmarketing.business.ajax.PermissionAjax.saveAssetPermissions(PermissionAjax.java:298)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.dotcms.repackage.org.directwebremoting.impl.CreatorModule$1.doFilter(CreatorModule.java:229)
	at com.dotcms.repackage.org.directwebremoting.impl.CreatorModule.executeMethod(CreatorModule.java:241)
	at com.dotcms.repackage.org.directwebremoting.impl.DefaultRemoter.execute(DefaultRemoter.java:379)
	at com.dotcms.repackage.org.directwebremoting.impl.DefaultRemoter.execute(DefaultRemoter.java:332)
```

### PR DESCRIPTION
**Pull request by freddyucv**
Avoid to run the ReindexQueueAPI#addStructureReindexEntries for each permission

### CODE REVIEW

File: dotCMS/src/main/java/com/dotmarketing/business/PermissionBitFactoryImpl.java [hunk 1]
```diff
@@ -1258,14 +1258,31 @@ private PersistResult persistPermission(Permission p) throws DotDataException {
 	@Override
     @Deprecated
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotmarketing/business/PermissionBitFactoryImpl.java:1259]: if it is deprecated please doc, which one is the new method or the suggestion to replace

File: dotCMS/src/main/java/com/dotmarketing/business/PermissionBitFactoryImpl.java [hunk 2]
```diff
@@ -1258,14 +1258,31 @@ private PersistResult persistPermission(Permission p) throws DotDataException {
...
+		boolean anyNew = false;
 
-    for(Permission permission:permissions) {
-      savePermission(permission, permissionable);
-    }
+		for(Permission permission:permissions) {
+			final PermissionSaveResult permissionSaveResult = savePermission(permission,
+					permissionable, false);
+
+			if (permissionSaveResult.result == PersistResult.NEW) {
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotmarketing/business/PermissionBitFactoryImpl.java:1267]: it could be handle by thread context, but I guess it is ok

**Review by jdotcms — APPROVED** [review]

File: dotCMS/src/integration-test/java/com/dotmarketing/business/PermissionBitFactoryImplTest.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+public class PermissionBitFactoryImplTest {
+
+    @Before
+    public void init() throws Exception {
+        IntegrationTestInitService.getInstance().init();
+    }
+
+    @Test
+    public void assignPermissions() throws DotDataException {
```
[the remark this answers is not in the document]
Comment (freddyucv) [reply dotCMS/src/integration-test/java/com/dotmarketing/business/PermissionBitFactoryImplTest.java:39]: done https://github.com/dotCMS/core/pull/21780/commits/c2ceed6de728d4fdb5001ed39d2f3760736a42b5#diff-ece4da8c7662efaefca2b50ae519ec6413720c7e8fc5934e1bd5a022528b33e3R31

File: dotCMS/src/main/java/com/dotmarketing/business/PermissionBitFactoryImpl.java (hunk 1 above)
Comment (freddyucv) [reply dotCMS/src/main/java/com/dotmarketing/business/PermissionBitFactoryImpl.java:1259]: done https://github.com/dotCMS/core/pull/21780/commits/c2ceed6de728d4fdb5001ed39d2f3760736a42b5#diff-e748015d1e1cc6e815fbfafdac7576623968742461c5525522cf3f9db77f8c29R1255

**Review by dsilvam — APPROVED** [review]

**Pull request merged**